### PR TITLE
 Adapted Phase2 Tracker Digitizer code to C++11 standard 

### PR DESCRIPTION
**Pull request by suchandradutta**
#### PR description:
overall code optimization which includes 
- Optimization of loops over simhits with std::copy_if
- Use of std::make_unique (C++14)
- STL loop improvement
- Clean-up of unnecessary header files, temporary variables, superfluous code
- Style uniformity

#### PR validation:
Digis validated using FourFourMuon (Pt: 10-100 GeV) + 200 PU events in CMSSW_11_1_X
( workflow number 20610.0 )  against the released version

Details can be found in the presentation : https://indico.cern.ch/event/838963/contributions/3679018/attachments/1964986/3267155/Digitizer_Code_Review_19Dec19.pdf

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)
@subirsarkar @emiglior

### CODE REVIEW

**Comment by kpedro88** [comment]
@suchandradutta please apply the code-checks and code-format, otherwise the PR tests will not start.
Also, please update the PR title to make it clear that it refers to the Phase 2 tracker digitization code.

**Comment by kpedro88** [comment]
please test

**Comment by civanch** [comment]
+1

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc [hunk 1]
```diff
@@ -116,20 +97,20 @@ namespace cms {
...
       for (auto const& det_u : pDD_->detUnits()) {
-        unsigned int detId_raw = det_u->geographicalId().rawId();
-        DetId detId = DetId(detId_raw);
-        if (DetId(detId).det() == DetId::Detector::Tracker) {
+        uint32_t rawId = det_u->geographicalId().rawId();
+        if (DetId(rawId).det() == DetId::Detector::Tracker) {
           const Phase2TrackerGeomDetUnit* pixdet = dynamic_cast<const Phase2TrackerGeomDetUnit*>(det_u);
           assert(pixdet);
-          detectorUnits_.insert(std::make_pair(detId_raw, pixdet));
+          detectorUnits_.insert({rawId, pixdet});
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc]: better to use `emplace(rawId,pixdet)`

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc [hunk 2]
```diff
@@ -323,15 +306,14 @@ namespace cms {
     }
 
     // Step C: create collection with the cache vector of DetSet
-    std::unique_ptr<edm::DetSetVector<PixelDigi> > output(new edm::DetSetVector<PixelDigi>(digiVector));
-    std::unique_ptr<edm::DetSetVector<PixelDigiSimLink> > outputlink(
-        new edm::DetSetVector<PixelDigiSimLink>(digiLinkVector));
+    std::unique_ptr<edm::DetSetVector<PixelDigi> > output = std::make_unique<edm::DetSetVector<PixelDigi> >(digiVector);
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc]: to reduce type repetition, you can write these as:
```
auto output = std::make_unique<edm::DetSetVector<PixelDigi>>(digiVector);
```
etc.

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc [hunk 3]
```diff
@@ -340,34 +322,36 @@ namespace {
     // so that when the row and column are inserted to PixelDigi the
     // coded channel stays the same (so that it can then be decoded
     // with Phase2TrackerDigi in stage2).
-    std::pair<int, int> ip = PixelDigi::channelToPixel(channel);
+    auto ip = PixelDigi::channelToPixel(channel);
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizer.cc]: `const auto&` (also below)

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 4]
```diff
@@ -26,36 +22,29 @@
...
 #include "Geometry/CommonTopologies/interface/PixelTopology.h"
+#include "Geometry/CommonDetUnit/interface/PixelGeomDetUnit.h"
 
-#include "CondFormats/SiPixelObjects/interface/PixelROC.h"
 using namespace edm;
 using namespace sipixelobjects;
 
+namespace {
+  // Mass in MeV
+  constexpr double m_pion = 139.571;
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: we may want to consider a common source within CMSSW for these mass values (other values are used elsewhere in some cases)

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 5]
```diff
@@ -345,69 +321,67 @@ void Phase2TrackerDigitizerAlgorithm::drift(const PSimHit& hit,
...
-  float DriftDistance;  // Distance between charge generation and collection
-  float DriftLength;    // Actual Drift Lentgh
-  float Sigma;
-
-  for (unsigned int i = 0; i != ionization_points.size(); ++i) {
-    float SegX, SegY, SegZ;  // position
-    SegX = ionization_points[i].x();
-    SegY = ionization_points[i].y();
-    SegZ = ionization_points[i].z();
+  //float Sigma_x = 1.;  // Charge spread
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: delete commented-out code

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 6]
```diff
@@ -345,69 +321,67 @@ void Phase2TrackerDigitizerAlgorithm::drift(const PSimHit& hit,
...
+    float Sigma_x = Sigma / CosLorenzAngleX;
+    float Sigma_y = Sigma / CosLorenzAngleY;
 
     // Insert a charge loss due to Rad Damage here
-    float energyOnCollector = ionization_points[i].energy();  // The energy that reaches the collector
+    float energyOnCollector = val.energy();  // The energy that reaches the collector
 
     // pseudoRadDamage
-    if (pseudoRadDamage >= 0.001) {
+    if (pseudoRadDamage_ >= 0.001) {
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: magic number 0.001 should be named constant

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 7]
```diff
@@ -495,82 +458,78 @@ void Phase2TrackerDigitizerAlgorithm::induce_signal(
...
-      else {
-        mp = MeasurementPoint(float(ix + 1), 0.0);
+      } else {
+        mp = MeasurementPoint(ix + 1, 0.0);
         xUB = topol->localPosition(mp).x();
         UpperBound = 1. - calcQ((xUB - CloudCenterX) / SigmaX);
       }
       float TotalIntegrationRange = UpperBound - LowerBound;  // get strip
-      x[ix] = TotalIntegrationRange;                          // save strip integral
+      x.insert({ix, TotalIntegrationRange});                  // save strip integral
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: use emplace

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 8]
```diff
@@ -495,82 +458,78 @@ void Phase2TrackerDigitizerAlgorithm::induce_signal(
...
-        mp = MeasurementPoint(0.0, float(iy + 1));
+      } else {
+        mp = MeasurementPoint(0.0, iy + 1);
         yUB = topol->localPosition(mp).y();
         UpperBound = 1. - calcQ((yUB - CloudCenterY) / SigmaY);
       }
 
       float TotalIntegrationRange = UpperBound - LowerBound;
-      y[iy] = TotalIntegrationRange;  // save strip integral
+      y.insert({iy, TotalIntegrationRange});  // save strip integral
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: use emplace

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 9]
```diff
@@ -889,20 +833,18 @@ void Phase2TrackerDigitizerAlgorithm::module_killing_DB(uint32_t detID) {
...
-  } else {  // all other module types: half-modules and single ROCs.
+  } else {               // all other module types: half-modules and single ROCs.
     // Get Bad ROC position:
     // follow the example of getBadRocPositions in CondFormats/SiPixelObjects/src/SiPixelQuality.cc
-    std::vector<GlobalPixel> badrocpositions(0);
-    for (unsigned int j = 0; j < 16; j++) {
-      if (SiPixelBadModule_->IsRocBad(detID, j) == true) {
-        std::vector<CablingPathToDetUnit> path = map_.product()->pathToDetUnit(detID);
+    std::vector<GlobalPixel> badrocpositions;
+    for (size_t j = 0; j < 16; j++) {
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: magic number 16 should be named constant

**Review by kpedro88 — COMMENTED** [review]
Thanks for these useful code improvements. In addition to the minor review comments, I have a few general comments:
* PSSDigitizerAlgorithm and PSPDigitizerAlgorithm appear to be the exact same code, just with a different name. I recommend making a generic PSDigitizerAlgorithm, and if you want to specify the PSS vs. PSP name for debugging etc., make that an input parameter of the algorithm.
* In general, we recommend that plugins just have a .cc file, with no separate .h file (since plugin code is not supposed to be included in other code).

**Comment by emiglior** [comment]
@skinnari FYI

**Comment by suchandradutta** [comment]
Incorporated suggestion for minor changes. 

About the general comments
  

-   **PSSDigitizerAlgorithm and PSPDigitizerAlgorithm appear to be the exact same code, just with a different name. I recommend making a generic PSDigitizerAlgorithm, and if you want to specify the PSS vs. PSP name for debugging etc., make that an input parameter of the algorithm.**

==> We expect different implementations following individual signal shape parametrization of PSP and PSS algorithms. The noise treatment may also differ. Once the picture becomes clear we shall take proper action.

-  **In general, we recommend that plugins just have a .cc file, with no separate .h file (since plugin code is not supposed to be included in other code).** 

Header files for Individual algorithms are merged in the .cc and removed. plugins/Phase2TrackerDigitizer.h is used by plugins/PreMixingPhase2TrackerWorker.cc. Similarly, plugins/Phase2TrackerDigitizerAlgorithm.h is used by individual algorithms.

**Comment by kpedro88** [comment]
please test

File: SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc [hunk 10]
```diff
@@ -345,69 +321,65 @@ void Phase2TrackerDigitizerAlgorithm::drift(const PSimHit& hit,
...
+    float Sigma_x = Sigma / CosLorenzAngleX;
+    float Sigma_y = Sigma / CosLorenzAngleY;
 
     // Insert a charge loss due to Rad Damage here
-    float energyOnCollector = ionization_points[i].energy();  // The energy that reaches the collector
+    float energyOnCollector = val.energy();  // The energy that reaches the collector
 
     // pseudoRadDamage
-    if (pseudoRadDamage >= 0.001) {
+    if (pseudoRadDamage_) {
```
Comment (kpedro88) [inline SimTracker/SiPhase2Digitizer/plugins/Phase2TrackerDigitizerAlgorithm.cc]: Is this change intentional? The behavior is different from the previous version of this line.

**Comment by suchandradutta** [comment]
Prior to the last commit I suspect I had an un-cleaned build  and the compilation error did not show up! Sorry for that!

**Comment by kpedro88** [comment]
please test

**Comment by silviodonato** [comment]
CMSSW_11_1_X will be used to run a rereco of GEN-SIM-DIGI-RAW produced with CMSSW_11_0_X for the HLT TDR.
This PR will not break the compatibility because it does not change the data format of the tracker digi. 
@suchandradutta could you confirm this PR will not break this compatibility?

**Comment by suchandradutta** [comment]
Yes I can confirm that this PR will not break the compatibility

**Comment by silviodonato** [comment]
Kind reminder of this PR to
upgrade:   @kpedro88 
simulation:   @mdhildreth  @civanch

**Comment by kpedro88** [comment]
+upgrade

**Comment by civanch** [comment]
+1

**Comment by silviodonato** [comment]
+1
